### ISSUE TITLE
200OK Blank Page

### ISSUE THREAD

**Issue by msp1kpj** [Bilal-S/iis2tomcat#42]
We are getting 200OK page back that are blank. If there was an error is suppose to through an exception?

The TomcatConnectErrorURL does not seem to be working. We have turned on logging but don't see what is the cause for the blank page.

**Comment by Bilal-S** [comment]
There would not be a redirect when the we receive communication from Tomcat.
The Error redirect would only be run when Tomcat fails to respond.
You can use standard IIS error page and redirection in that case.

You have a few options to trace further if you believe this to be connector related:
a) Check Tomcat Logs for errors
b) You can turn on connector debug level logging (set log level to 4) 
c) Enable windows application level logging by connector: http://boncode.net/connector/webdocs/Tomcat_Connector.htm#_Toc470191887
d) Turn on IIS trace and check content of trace files
e) Use a package capture tool like wireshark and capture traffic on HTTP port as well as AJP port to see whether they correspond


I would work from a - e to see whether this is a conenctor error somehow.
However, normally if transmission occurs from tomcat to iis with HTTP 200, everything is working as it should. I would think looking at tomcat first to change the behavior to what you expect would probably be most advisable.

**Comment by tkiansoon** [comment]
Hi Bilal,

I'm continuing the issue originally posted, we have turn on logging on Windows Application level, connector debug level logging (headers as well since hard to parse through debug level logging). 

I having two scenario and questions:
1. We stopped tomcat service to ensure the logging works and we do see Windows event log (event id = 417) but it still returning 200 OK status, shouldn't this return 500 or some exception error status code? I can't tell my LB not to send traffic if the app server when down.

2. When we are receiving file length=0, it doesn't log into Windows event (event id = 417) as we expecting, therefore it doesn't help very much. 

Is there any sort of Unique Identifier to tie the log by request since there log are scrambled each other and hard to see which line ties to which requests?

We still trying turning on Tomcat logging (somehow it overwite Server.xml everytime we restart the service) to see if we can get useful information, but I do need the unique identifier from Boncode as reference key with tomcat if possible.

**Comment by tkiansoon** [comment]
Also, we make sure the packetSize and the compression issue wasn't the case.

**Comment by Bilal-S** [comment]
This sounds like you have the `EnableHTTPStatusCodes` setting set to false. Can you provide the BonCodeAJP13.settings file or change the setting in there to true. This will send the HTTP status from tomcat along.

You can run a connector command like this one to find out which setting file is in use:

`http://localhost/a.jsp?BonCodeConnectorVersion=true`

Also tomcat server.xml files should not be rewritten every time your tomcat server starts. This does not appear to be expected behavior.

**Comment by tkiansoon** [comment]
1. EnableHTTPStatusCodes wasn't set but we explicitly set to True and we still getting 200 OK status with file length=0. Can you help me understand why when we stop the Tomcat server, it still reply 200 OK?

2. I apologize it was ROOT.xml, not server.xml for Tomcat. 

Attached is the BonCodeAJP13 settings:
<Settings>
    <Server>x.x.x.x</Server>
    <Port>xxxxx</Port>
    <EnableHeaderDataSupport>True</EnableHeaderDataSupport>
	<EnableHTTPStatusCodes>True</EnableHTTPStatusCodes>
	<ForceSecureSession>False</ForceSecureSession>
	<AllowEmptyHeaders>False</AllowEmptyHeaders>
	<DocRoot>/my/path/to/wwwroot/</DocRoot>
	<TomcatConnectErrorURL>/probes/TomcatErrorMessage.aspx</TomcatConnectErrorURL>
    <SupportEmail>email@example.com</SupportEmail>
    <SiteName>www.mywebsite.com</SiteName>
    <PacketSize>65536</PacketSize>
    <MaxConnections>500</MaxConnections>
    <LogLevel>0</LogLevel>
    <LogDir>\\logDir\</LogDir>
</Settings>

**Comment by tkiansoon** [comment]
`<Settings>
    <Server>x.x.x.x</Server>
    <Port>xxxxx</Port>
    <EnableHeaderDataSupport>True</EnableHeaderDataSupport>
	<EnableHTTPStatusCodes>True</EnableHTTPStatusCodes>
	<ForceSecureSession>False</ForceSecureSession>
	<AllowEmptyHeaders>False</AllowEmptyHeaders>
	<DocRoot>/my/path/to/wwwroot/</DocRoot>
	<TomcatConnectErrorURL>/probes/TomcatErrorMessage.aspx</TomcatConnectErrorURL>
    <SupportEmail>email@example.com</SupportEmail>
    <SiteName>www.mywebsite.com</SiteName>
    <PacketSize>65536</PacketSize>
    <MaxConnections>500</MaxConnections>
    <LogLevel>0</LogLevel>
    <LogDir>\\logDir\</LogDir>
</Settings>
`

**Comment by Bilal-S** [comment]
Thanks for checking.
If EnableHTTPStatusCodes was not set and you are using the correct Boncode config file the only other scenario would be that Tomcat returns 200 codes to you that are being passed on.
The connector would only mirror what is provided in that case.

Are you using tomcat as app server directly or using some other package such as Railo or Lucee as intermediary? Those intermediaries also have an HTTP code override.

If you are using directly have you captured any of the AJP traffic? This could show you whether the is an actual return from tomcat besides the status code?

**Comment by msp1kpj** [comment]
Bilal,

We were able to dig into this further and this is what we found.

The TomcatConnectErrorURL setting will only display if you are on the local server. This is different form the documentation. If you are not on the local server and are using the TomcatConnectErrorURL it will return a status code of 200OK with no page context. 

The CallHandler.cs file the method PrintError uses a IsLocalIP method to determine if it will display the Connection Error or custom connection url.

**Comment by Bilal-S** [comment]
OK.
I am getting a hint here what could go on.
So it is correct that a error detail will only be shown to user when the request is run on server if there is a connection problem. A more generic error is displayed for any remote invocation: 
```
Generic Connector Communication Error: <hr>Please check and adjust your setup:<br>Ensure that Tomcat is running on given host and port.<br>If this is a timeout error consider adjusting IIS timeout by changing executionTimeout attribute in web.config (see manual).
```
Is that what you see?
If not, are you redirecting?  To where exactly? And, how do you handle errors? In short, your redirection page has to ensure that it handles the connection error. 

So things of interest:
a) What is the error you are getting shown when you run this locally without redirection?
b) What does your redirect page do?

**Comment by Bilal-S** [comment]
OK. 
I see what you mean. The redirect is disabled if not local. I will enable it again.

**Comment by Bilal-S** [comment]
Please check release 1.0.34  and let me know.
This has nothing todo with 200 OKs that is still something you will need to resolve on your redirects. However, the redirect should work for remote clients as well.
https://github.com/Bilal-S/iis2tomcat/releases/tag/1.0.34

**Comment by tkiansoon** [comment]
Hi Bilal, the update looks good except couple things:
1. HttpUtility.UrlPathEncode() doesn't encode all the html tags, so we change to use:
strPublicErr = HttpUtility.UrlEncode(HttpUtility.HtmlEncode(strPublicErr.Substring(0, Math.Min(strPublicErr.Length, 1199))));

2. Is it possible to log the strPublicErr using RecordSysEvent() we get the detail of the error since you cut the detail off from the strFullUrl if no local request?

We move to next step with our errors and would like to have your input if you don't mind. Somehow we got Tomcat connection reset error that causing the Boncode error, this is the error detail:

Generic Connector Communication Error: 
Please check and adjust your setup:
Ensure that Tomcat is running on given host and port. If this is a timeout error consider adjusting IIS timeout by changing executionTimeout attribute in web.config (see manual).-Thread was being aborted.    at System.Threading.Thread.AbortInternal()
   at System.Threading.Thread.Abort(Object stateInfo)
   at System.Web.HttpResponse.AbortCurrentThread()
   at BonCodeIIS.BonCodeCallHandler.PrintError(HttpContext context, String strPublicErr, String strLocalErr)
   at BonCodeIIS.BonCodeCallHandler.ProcessRequest(HttpContext context)

**Comment by Bilal-S** [comment]
@tkiansoon the detail error should be supported in 1.0.35. Would be helpful to see the error detail that leads to internal abort.

**Comment by tkiansoon** [comment]
We removed the <PacketSize>65536<PacketSize> and use the default and that solve the drop connections issue. Along with this setting, if we remove <MaxConnections>, IIS CPU will max'ed out and requires IIS service restart for the site. So we set <MaxConnections>200</MaxConnections> and that helps a little, still has drop connections. It's not a solution but at least it buy us time to troubleshoot/workaround.

**Comment by Bilal-S** [comment]
Do you have a minimal test case that you could share with me?

**Comment by Bilal-S** [comment]
closing for now.

**Comment by Bilal-S** [comment]
Please also try this in your server.xml for Tomcat for the AJP protocol definition:
```
connectionTimeout="60000" keepAliveTimeout="-1"
```
